From c08ba71eba8db6a70037329dd989ba8bd1eeb653 Mon Sep 17 00:00:00 2001
From: Vinh Nguyen Trong <Vinh.NguyenTrong@vn.bosch.com>
Date: Mon, 22 Apr 2024 19:30:29 +0700
Subject: [PATCH 2/3] drm: rcar-du: cmm: use atomic for vsync instead of spin
 lock

To reduce consuming time when the race between interrupt handler
(e.g. rcar_du_cmm_kick()) and the worker thread, the atomic variable
can be used for vsync.

Signed-off-by: Vinh Nguyen Trong <Vinh.NguyenTrong@vn.bosch.com>
Signed-off-by: Phong Hoang <phong.hoang.wz@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_cmm.c | 23 +++++------------------
 1 file changed, 5 insertions(+), 18 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_cmm.c b/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
index a42a88e19cc8..7c33dbc0f774 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_cmm.c
@@ -148,12 +148,12 @@ struct rcar_du_cmm {
 	struct work_struct work;
 
 	struct cmm_reg_save reg_save;
+	atomic_t vsync;
 	bool active;
 	bool dbuf;
 	bool clu_dbuf;
 	bool init;
 	bool direct;
-	bool vsync;
 	bool authority;
 	pid_t pid;
 	bool soc_support;
@@ -167,7 +167,6 @@ struct rcar_du_cmm_file_priv {
 };
 
 static DEFINE_MUTEX(cmm_event_lock);
-static DEFINE_SPINLOCK(cmm_direct_lock);
 
 static inline void event_prev_cancel_locked(struct cmm_module_t *module);
 
@@ -1355,24 +1354,12 @@ static int hgo_get(struct rcar_du_cmm *du_cmm,
 
 static bool du_cmm_vsync_get(struct rcar_du_cmm *du_cmm)
 {
-	unsigned long flags;
-	bool vsync;
-
-	spin_lock_irqsave(&cmm_direct_lock, flags);
-	vsync = du_cmm->vsync;
-	du_cmm->vsync = false;
-	spin_unlock_irqrestore(&cmm_direct_lock, flags);
-
-	return vsync;
+	return atomic_xchg(&du_cmm->vsync, 0);
 }
 
-static void du_cmm_vsync_set(struct rcar_du_cmm *du_cmm, bool vsync)
+static void du_cmm_vsync_set(struct rcar_du_cmm *du_cmm)
 {
-	unsigned long flags;
-
-	spin_lock_irqsave(&cmm_direct_lock, flags);
-	du_cmm->vsync = vsync;
-	spin_unlock_irqrestore(&cmm_direct_lock, flags);
+	atomic_set(&du_cmm->vsync, 1);
 }
 
 static void du_cmm_work(struct work_struct *work)
@@ -1473,7 +1460,7 @@ void rcar_du_cmm_kick(struct rcar_du_crtc *rcrtc)
 		return;
 
 	if (!du_cmm_que_empty(du_cmm)) {
-		du_cmm_vsync_set(du_cmm, true);
+		du_cmm_vsync_set(du_cmm);
 		queue_work(du_cmm->workqueue, &du_cmm->work);
 	}
 }
-- 
2.25.1

